### ISSUE TITLE
Test what you see

### ISSUE THREAD

**Issue by andreareginato** [andreareginato/betterspecs#14]
Write your thoughts about the "[test what you see](http://betterspecs.org/#integration)" best practice.

**Comment by onebree** [comment]
Right now, we are at ~70% coverage and rely only on, what I guess to be, integration testing. We use Capybara to test the gui as if an admin/user were clicking through it. We "know" the data successfully saved to the database when it renders within a table. I agree that it is slow. It can also be annoying as Capybara may not find an element, but I can clearly see/use it in development mode.

Any/all validations are presented as success/error messages via our models (sometimes controllers if complex). We then check to make sure the error messages can be found by Capybara.

I also agree that if the UI changes, just as ours just did moving to bootstrap, it requires a complete overhaul of the specs and views.

How would you suggest we manage our testing, then? Continue as integration and change whenever a label/view changes (not too often, manageable), or separate into MVC/routes/etc specs?

**Comment by marnen** [comment]
@onebree Your Capybara tests sound OK as far as they go (and I hope you're driving them from Cucumber). But do add some unit specs for the models too -- there are some things that are impractical to test solely from integration tests. (And get into the habit of developing story- and test-first!)

The specs changing if the UI changes is a *feature*, not a bug, at least to some extent, but there are ways of using helper functions or (if you're not using Cucumber) page objects to make this more manageable.

**Comment by onebree** [comment]
One of our Capybara tests goes as follows:
```ruby
context "HAPPY PATHS" do
  #...
  scenario "create a parking lot and check hints count" do
    @tenant.default_location.parking_lots.first.destroy
    click_link "Create Parking Lot"
    fill_in "Name", with: "Parking Lot 4"
    select "Corporate", from: "Workgroup"
    click_button "Submit"
    success_messages ["Parking Lot successfully created."]
    has_no_error_message
    expect(page).to have_css("table#parking_lots tr", count: 3)
    expect(page).to have_css("table#parking_lots tr", text: "Parking Lot 4")
    expect(@tenant.default_location.parking_lots.find_by_name("Parking Lot 4").hints.count).to eq(4)
  end
  # ...
end
```

If we need to test dependency validations, we create or destroy the dependency immediately before using Capybara to find gui elements.

**Comment by marnen** [comment]
I'm not sure what you're attempting to describe to me or get my input on here.

I am sure, though, that you should be using Cucumber for this. Always describe your UI in user terms, not programmer terms.

**Comment by onebree** [comment]
Okay. I wanted to ask if the type of tests we write are _enough_. I do agree that Cucumber is better for testing the GUI, but I was not here yet when they decided. The reason they chose RSpec, I guess, is to have the ability to test models, such as in the last assertion. They also decided to _only_ test the gui through `features` specs, as it seemed to be the most important thing to test.

Basically all tests were written _after_ MVC. I was just trying to get help from this discussion as it _is_ about testing what you see/integration testing.

**Comment by marnen** [comment]
> Okay. I wanted to ask if the type of tests we write are *enough*. 

I'd say it's unlikely, but not absolutely impossible. How do you test complex model logic?

> The reason they chose RSpec, I guess, is to have the ability to test models, such as in the last assertion.

Then clearly, they didn't know enough to make a useful decision: Cucumber steps are generally implemented in RSpec, so that would have been no problem.

> They also decided to *only* test the gui through `features` specs, as it seemed to be the most important thing to test.

Yes, if you have to only have one kind of test for some reason, that is generally the right kind, because it *can* test everything, whereas unit tests cannot. But unit tests are often useful anyway; it's impractical to set up an integration test for every permutation of complex model logic.

> Basically all tests were written *after* MVC.

...which could explain the abysmal coverage numbers. Are they at least doing all new development test-first?

**Comment by Fryie** [comment]
"Test what you see" seems odd to me. In fact, users don't "see" your models either. Yet still you are not advocating that we only write acceptance tests?

The most important problems with acceptance tests are:
- they don't scale (I cannot acceptance-test EVERY path through my code)
- they are slow (even if Rails controller tests are slow themselves, they're still slower than an acceptance test - especially on a JS-heavy website)

Of course I write acceptance tests for every user story. But what about error handling? What about "my redis connection died and I cannot enqueue this job"? What about "the third-party API is down and I cannot process payments"? What about invalid or malicious user input, etc.? Testing all of these cases (which may well appear on multiple parts of the site) through the UI doesn't seem practical to me.

If you say that there is no value in unit testing controllers you're basically saying that controllers are not units within your code (either that, or you're against unit testing in general).

The controller's responsibility is a) to take input from the user and pass it to the app and b) to parse the response from the app and deal with it in a way that makes sense to a user (either by returning some JSON in an API or by setting up instance variables for a view). That's enough responsibility that a unit test makes sense at least in some cases.

Also, on another note, "skinny controller, fat model" is an antipattern IMHO. The more reasonable approach is "skinny everything". And just because classes are skinny, doesn't mean there's no value in unit testing them.

To each their own, I guess, but I see no a priori reason why unit testing controllers should considered to be bad.

**Comment by marnen** [comment]
@Fryie:

> "Test what you see" seems odd to me. 

Why? It is the only way to know that your application works as expected for the user. 

> In fact, users don't "see" your models either. Yet still you are not advocating that we only write acceptance tests?

Acceptance tests are the most important tests, because they are the only tests that can show that the application is working as a whole.

I write unit tests to make refactoring easier and to increase the number of execution paths I can test, not because I think they will tell me anything about whether the application as a whole works properly. Unit tests -- even controller unit tests -- cannot give you any assurance at all that the application works as desired for the user. Only acceptance tests can do that. 

> The most important problems with acceptance tests are:

> they don't scale (I cannot acceptance-test EVERY path through my code)

This right here is a big reason to have unit tests. The other big reason is that it's often more efficient to test at a smaller granularity before running acceptance tests. 

However, don't delude yourself into thinking that you can feasibly unit-test every path through your code either. That quickly runs into huge combinatorial explosions. C0 plus some common C1 use cases is usually as good as you're going to get. 

> they are slow (even if Rails controller tests are slow themselves, they're still slower than an acceptance test - especially on a JS-heavy website)

They can actually be quite fast if you use a headless browser.  But I'd rather my tests be correct than fast. 

> Of course I write acceptance tests for every user story. But what about error handling? What about "my redis connection died and I cannot enqueue this job"? What about "the third-party API is down and I cannot process payments"? 

Acceptance tests are ideal for this. Utilities such as WebMock are great for simulating error conditions caused by external services. 

> What about invalid or malicious user input, etc.? 

You can use acceptance tests for that. 

> Testing all of these cases (which may well appear on multiple parts of the site) through the UI doesn't seem practical to me.

It's quite practical. I do it on a regular basis. Read my earlier posts in this thread for suggestions on how. 

Any test that simulates user input should involve the UI, I think. 

> If you say that there is no value in unit testing controllers you're basically saying that controllers are not meaningful units within your code (either that, or you're against unit testing in general).

I think you may have responded to this thread without actually reading it, because I already answered this above. Briefly put, while a controller may be a meaningful unit, in a proper Rails application, the controller should be nothing more than a tiny piece of brain-dead glue code. All the logic should be in models and service objects, so you can't meaningfully unit-test your controllers without mocking all their collaborators -- and at that point, you're testing your mocks, not your controllers. 

If your controllers have enough logic in them to benefit from unit tests, then you should refactor them. 

Am I against unit testing in general? No. But I don't think it's anywhere near as important as acceptance testing. I'd be willing to omit unit tests for simple models too, *if* I had suitable acceptance tests (though I don't tend to do this in practice). 

> The controller's responsibility is a) to take input from the user and pass it to the app

Yes.

> and b) to parse the response from the app and deal with it in a way that makes sense to a user

Not as you stated. Marshaling data for the view is the controller's responsibility, but the parsing and processing of user input is supposed to take place in the model layer. 

> That's enough responsibility that a unit test makes sense at least in some cases.

And that's what's wrong with your controllers. You should create model methods for the processing. Then the controller consumes the results of those methods and passes them to the view. 

> Also, on another note, "skinny controller, fat model" is an antipattern IMHO. The more reasonable approach is "skinny everything". 

"Fat model" as I interpret it means that the model *layer* is fat. Of course no one class should be fat. But almost nothing should happen in the controller layer. The controller just brainlessly passes data from model to view, while all the logic is in the model layer. 

That's what "skinny controller, fat model" really means, I believe. And that is proper Rails MVC, not an antipattern. 

> And just because classes are skinny, doesn't mean there's no value in unit testing them.

If a class does nothing but pass data back and forth between its collaborators, and contains no logic of its own, then there is nothing to unit test that's meaningful. 

And that is what a proper Rails controller should look like. 

> To each their own, I guess,

Software engineering ought to be based on demonstrable facts. I do not believe that "to each their own" is a phrase that has any place in this discipline. 

> but I see no a priori reason why unit testing controllers should considered to be bad.

Because it's unnecessary, it's painful to implement, it verifies the wrong things, and it's only useful if you're doing MVC wrong. Add those facts up and the conclusion is quite clear: it's a waste of time and effort. The time would be better spent refactoring logic into the model layer and writing acceptance tests.

**Comment by sdeframond** [comment]
Relevant: a [very interesting 
discussion](http://martinfowler.com/articles/is-tdd-dead/) between DHH, 
Martin Fowler and Kent Beck about what to test and why.

**Comment by marnen** [comment]
@sdeframond: ooh, that's fascinating (though I wish they'd transcribed it -- I have more time to read than to watch video)! I'll note, however, that DHH apparently thinks hexagonal Rails is a bad thing, while I think it's a path to better use of Rails MVC. (In general I don't think much of DHH these days -- he did a great job with early Rails, but can't seem to see beyond it.)

**Comment by Fryie** [comment]
> Software engineering ought to be based on demonstrable facts. I do not believe that "to each their own" is a phrase that has any place in this discipline.
Something does not become a science simply by someone stating that it is. If you want demonstrable facts, please cite extensive peer-reviewed studies, not just personal anecdotes with thousands of possible confounders. Failing that, I suppose that yes, it is indeed a matter of opinion.

I think we have different ways of writing both applications and tests in general. Most importantly, I do not write tests to verify correctness. In fact, in almost all cases I am much quicker verifying correct behaviour by hand.

The reason I test is to catch regressions, and for that purpose unit tests are as valuable as integration tests. Of course unit tests can be brittle, but so can integration tests ("whoops, I changed the post login message, now I have to rewrite my tests"), so it's all a question of writing good tests anyway.

> Not as you stated. Marshaling data for the view is the controller's responsibility, but the parsing and processing of user input is supposed to take place in the model layer.
I disagree. My models should not have to know that they are part of a Rails app (it's bad enough that they're tightly coupled to ActiveRecord, i.e. to the database). I should be able to take my models and reuse them somewhere else; that's not just theory, I've done that before.

The controller on the other hand is the boundary, the entity that translates web requests into business requests and business responses into web responses (including setting up instance variables for the view, rendering a JSON representation, etc). IMHO, this is similar to the approach discussed by Uncle Bob in [Architecture: The Lost Years](https://www.youtube.com/watch?v=WpkDN78P884).

**Comment by marnen** [comment]
@Fryie:

> Something does not become a science simply by someone stating that it is.

...which I didn't do. I merely said that we should take an evidence-based approach to software engineering, not that it was a science. 

> Most importantly, I do not write tests to verify correctness. In fact, in almost all cases I am much quicker verifying correct behaviour by hand.

What do you write tests for, then? (Yes, I see you said regressions, but is that the only purpose?) And what do you mean by "correctness" here? (I can think of at least two meanings that make sense in this context.)

You're right that it's generally quicker to run a one-off test by hand than to automate it. However, refactoring and other maintenance activities require repetition of the same tests over and over, to make sure that nothing breaks during the process. And that's where automation comes in. 

Also, think about this: the way you make sure your application is correct *is* by catching regressions. And catching regressions has no purpose except to ensure your application is correct. So what do you mean when you say you don't write tests to ensure that the application is correct? What would that kind of test be, to you?

> My models should not have to know that they are part of a Rails app

Agreed. That is completely orthogonal to the idea that the model layer should be where data processing happens. The model can process data received from the controller without knowing where it came from or where it's going -- and it should. The controller should call model methods and consume the return values. 

The service objects, OTOH (which are part of the model layer but somewhat different) generally probably *do* know that they're part of a Rails app, and have less reusability. 

> The controller on the other hand is the boundary, the entity that translates web requests into business requests and business responses into web responses

That is how Reenskaug-style MVC is designed, as used in Smalltalk and in Cocoa. It is emphatically *not* how Rails MVC (originally called Model2 MVC, from the name of an earlier implementation of this pattern) is intended to work. There is more than one MVC architecture paradigm out there, and the term is thus a bit overloaded. 

Reenskaug MVC doesn't work very well for server-side Web applications, or so I understand.  It's unfortunate that the related but different paradigm for Web applications is called by the same name even though it's not the same thing. 

The controller in Reenskaug MVC is an intelligent object, responsible for quite a lot of stuff (and thus it makes sense to unit-test it). The controller in Rails MVC should be as dumb as you can possibly make it, simply delegating to intelligent classes in the model layer. 

Right now it sounds like you're trying to do Reenskaug MVC in a framework that is designed for a different pattern altogether.

**Comment by sdeframond** [comment]
By the way, does any of you know how to do proper integration tests with 
asynchronous jobs? I am working with Sidekiq now but information about 
other frameworks would be interesting too.

I haven't found better than to test that 1) the job was correctly 
enqueued, then 2) the job does what it should when run. This is kind of 
a mixed approach between unit and integration testing. I whish I could 
write end-to-end integration tests though.

What do you think?

-Sam

**Comment by Fryie** [comment]
While there are options to inline sidekiq so the jobs are run directly (see [the docs](https://github.com/mperham/sidekiq/wiki/Testing)), I think it's not a particulary good idea.

Basically there are several (not exclusive, but complementary strategies):
- Test both components in isolation (the enqueuing and the worker itself) - which you already do
- Test the contract between the two (e.g. with something like [pacto](http://thoughtworks.github.io/pacto/))
- Have a dedicated test environment that loads both components and runs an end-to-end test

The last option is very expensive (and sometimes not feasible, e.g. if you have a delay for the worker), so IMHO it should be restricted to very important cases.
I think [this presentation](http://martinfowler.com/articles/microservice-testing/) by Martin Fowler about microservices testing is interesting in this regard.

**Comment by marnen** [comment]
@sdeframond:
> By the way, does any of you know how to do proper integration tests with
asynchronous jobs?

That's a really interesting question. I'm trying to remember how I've handled this in the past. 

I think I would tend to have the job run synchronously in the testing environment; otherwise there's really no way to follow the whole process through. Or (mostly equivalent) somehow get a promise from the enqueued job that resolves when it's done, and test that. 

It seems to me that this is the approach I've taken in the past: run synchronously if `Rails.env.test?`. But I don't have access to that code now to check. 

Another approach would be to treat the queue like an external service: mock the results of the enqueued job, then test the job logic elsewhere. 

I'm not sure which of these I like better. They all have problems. I'd tend to decide on a case-by-case basis.

**Comment by tgaff** [comment]
Considering how contentious this rule is; I really think it should be removed from the guide as a recommendation.  
If the goal of betterspecs is to suggest guidelines based on the collected knowledge, experience and general consensus of RSpec users (on the best ways to use RSpec); then it doesn't seem that we're serving that goal by listing something with so little agreement.

**Comment by marnen** [comment]
@tgaff The principle isn't controversial. You'll notice that most of the disagreement comes from people who are new to RSpec, not from experienced RSpec users. 

Besides, if it were obvious, there would be no need to recommend it, would there?

**Comment by Fryie** [comment]
I don't want to add to this discussion again, but to state that this principle is not controversial is ... [quite a statement](http://solnic.eu/2012/02/02/yes-you-should-write-controller-tests.html). I also just watched a screencast by Gary Bernhard yesterday, where he refactored a big controller test - in the end the controller spec was smaller and more isolated but it was still kept in place.

I'll furthermore [refer to this blog post](http://www.satisfice.com/blog/archives/27).

**Comment by ryanmtaylor** [comment]
Testing controllers as an integration test, before they form a UI is one thing.

Testing a controller (like an API) because the application has to interface with something is Contract Testing.

I think there's also a case to be made that testing create/update/delete within a controller is helpful when the integrity of two models is inter-connected, but I think this suggests that your updates should have an interface outside of your controller, and you should instead test that interface as a representation of the domain and business requirements (feature) of the app -- not the ins and outs of the controller itself.

**Comment by Fryie** [comment]
@ryanmtaylor Just answer me this question: what is a controller's responsibility? Because before you answer that question you can't answer whether and how controllers should be tested.

My opinion is that Rails controllers by design / common usage do a really, really poor job at adhering to the SRP. Are they supposed to cover business logic? Parameter handling? Directing routes to specific entrypoints?
Different projects will use controllers in different ways. If your controllers are extremely dumb (i.e. they do exactly zero things before piping everything into, say, a Trailblazer operation), then by all means don't unit test them. But that's probably not how 95% of Rails projects are set up. As soon as there is even just one strong parameters method in there, you basically have parameter validation, and then that's logic that should be covered in your tests.

**Comment by brunofacca** [comment]
IMO this guideline should be updated. 

> Now I only create integration tests using RSpec and Capybara.

It should probably say RSpec *and/or* Capybara.  I don't think feature specs with Capybara can replace controller specs and/or request specs in most cases. Even if they could, it would result in a slow and likely brittle a test suite.

For instance, Capybara is [not suited for testing APIs](https://www.varvet.com/blog/capybara-and-testing-apis/). You should probably mention [request specs](https://relishapp.com/rspec/rspec-rails/docs/request-specs/request-spec), which are suited for API testing and widely used in Rails 5 as a replacement for controller specs.

**Comment by marnen** [comment]
@Fryie:

> If your controllers are extremely dumb (i.e. they do exactly zero things before piping everything into, say, a Trailblazer operation), then by all means don't unit test them. 

That's how I strive to structure my controllers. Anything complex goes into a model or a service object (which is really a special kind of model).

> But that's probably not how 95% of Rails projects are set up.

You may be right, but "skinny controller, fat model" has long been the Rails ideal.

> As soon as there is even just one strong parameters method in there, you basically have parameter validation, and then that's logic that should be covered in your tests.

...which is why I hate strong parameters. That kind of logic belongs in the model layer. (As I've said elsewhere in this thread, I basically proxy my strong parameter logic to the model layer in my applications. That way I work with, rather than against, the framework, but still keep the data validation in my models.)

**Comment by brunofacca** [comment]
@marnen:

> They can (especially with Cucumber on top) and should, if the controllers are sufficiently skinny. (If they're not sufficiently skinny, that's a problem to fix, not to palliate with more tests.)

I'm referring to applications with extremely skinny controllers. Let me rephrase: Feature specs with Capybara may replace controller/request specs, but it will result in a slower test suite. Additionally, finding the problems behind test failures should take longer (lower-level tests help to pinpoint issues) and the fact that your tests rely on DOM elements will likely increase brittleness.

> For API testing, I very highly recommend the https://github.com/zipmark/rspec_api_documentation gem

I did not know that gem. Looks great, thank you for the recommendation.

**Comment by marnen** [comment]
@brunofacca:

> Let me rephrase: Feature specs with Capybara may replace controller/request specs, but it will result in a slower test suite.

So what? The tests will be more correct, _especially_ if the controllers are skinny.

> finding the problems behind test failures should take longer (lower-level tests help to pinpoint issues)

That's why we have model unit tests. If the controllers are sufficiently skinny, there is essentially nothing in them that will break, so there is nothing in them to test.

> and the fact that your tests rely on DOM elements will likely increase brittleness

The fact that the tests rely on DOM elements increases _correctness_. I _want_ my tests to break when the DOM changes significantly, because the DOM is what the user interacts with.

**Comment by brunofacca** [comment]
> So what? The tests will be more correct, especially if the controllers are skinny.
@marnen:

IMO you can't enjoy the full benefits of the TDD process with a slow feedback loop. Also, if tests take too long to run, it is easier to loose focus. I believe tests can be correct and fast at the same time.

> The fact that the tests rely on DOM elements increases correctness. I want my tests to break when the DOM changes significantly, because the DOM is what the user interacts with.

The DOM may not have to change significantly to break your tests. 

Anyway, I suggest we agree to disagree, as this seems like one of those discussions that can go on forever.

**Comment by marnen** [comment]
@brunofacca:
> IMO you can't enjoy the full benefits of the TDD process with a slow feedback loop.

That's true to a point, which is why we use the model unit tests as the inner loop. Then we use the feature tests as an outer loop if they're too slow to run all the time.

Personally, though, I am so convinced of the benefits of test-first development that I'm willing to be pretty patient for slightly slower _but correct_ tests to run.

> Also, if tests take too long to run, it is easier to loose focus. 

I use Guard and other tools that help track that focus for me.

> I believe tests can be correct and fast at the same time.

That's the ideal, but if I have to choose one of the two, I will choose correct over fast nearly every time—within reason: I don't recompile the Ruby interpreter for every test, after all. :)

> Anyway, I suggest we agree to disagree

I don't do that; I think it's sloppy. This is a field where facts are verifiable, which means that if we disagree, then probably at least one (more likely both) of us has more to learn.

**Comment by brunofacca** [comment]
@marnen:

> I don't do that; I think it's sloppy. This is a field where facts are verifiable, which means that if we disagree, then probably at least one (more likely both) of us has more to learn.

The fact that facts are verifiable in this field does not stop some opinion-based discussions from going on "forever", such as: a) The mockists vs. classicists discussion; b) Extracting model code to concerns vs. service objects in Rails (DHH defends the former). I believe our attachment to specific points of view and the fact that most people don't like being proven wrong (even though it provides a great opportunity for learning) tends to skew our perception/interpretation of verifiable facts. The fact that there is always some degree of subjectivity involved makes it hard for us to see things objectively (if that's even possible). I believe they call it "confirmation bias". Anyway, that's a whole other (philosophical) discussion. 

I agree that, as long as we enter this kind of technical "debate" with an open mind, we can learn a lot. I sure learned from out discussion, as I'm relatively new to testing and striving to learn as much as I can about the "best practices". In that spirit, thanks for a good talk :)

**Comment by brunofacca** [comment]
@marnen: I had given some thought to the points discussed in the previous comments and tried out your approach while developing a small Rails app over the last few weeks. I did not write any controller tests for web views. I have changed my mind and now agree that feature specs can replace controller specs in web apps if the controllers are skinny enough (so if a feature spec fails due to a controller problem, the cause would be quick to pinpoint). After all, everything that we may test in a controller spec is also tested in a well-written feature spec. Although feature specs are much slower than controller specs, our apps need them for UI testing anyway, so the slowness is inevitable.

**Comment by marnen** [comment]
@brunofacca:
> Although feature specs are much slower than controller specs, our apps need them for UI testing anyway, so the slowness is inevitable.

That's basically my view. I'm glad it worked out in practice for you!

**Comment by aesyondu** [comment]
> I don't want to add to this discussion again, but to state that this principle is not controversial is ... quite a statement. I also just watched a screencast by Gary Bernhard yesterday, where he refactored a big controller test - in the end the controller spec was smaller and more isolated but it was still kept in place.

Not sure if it's just me, but the link provided now resolves to a porn site, `http://solnic.eu/.*`. So be WARNED.

**Comment by marnen** [comment]
@aesyondu Yikes! Looks like Piotr Solnica's site is now at http://solnic.codes, for what it's worth.
